### PR TITLE
Fix compiler crash on assignment to a narrowed variable in a query action

### LINKED ISSUE #32271: Compiler crashes on assignment to a narrowed variable in a query action
**Description:**
Should result in a compilation error (because the variable is implicitly final).

**Steps to reproduce:**
```ballerina
public function main() {
    (int|string)[] x = [1, 2];

    from var a in x 
        where a is int 
        do {
            a = "hello";
        };
}
```

```log
[2021-08-22 22:14:44,068] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolResolver.lookupSymbol(SymbolResolver.java:591)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolResolver.lookupSymbolInMainSpace(SymbolResolver.java:620)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.defineOriginalSymbol(SemanticAnalyzer.java:4368)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.defineOriginalSymbol(SemanticAnalyzer.java:4382)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.defineOriginalSymbol(SemanticAnalyzer.java:4382)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.defineOriginalSymbol(SemanticAnalyzer.java:4382)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.defineOriginalSymbol(SemanticAnalyzer.java:4382)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.defineOriginalSymbol(SemanticAnalyzer.java:4382)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.resetTypeNarrowing(SemanticAnalyzer.java:4363)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1861)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangAssignment.accept(BLangAssignment.java:80)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.lambda$visit$28(SemanticAnalyzer.java:1739)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1739)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:59)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:5141)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangQueryAction.accept(BLangQueryAction.java:41)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:374)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:353)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:2323)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangExpressionStmt.accept(BLangExpressionStmt.java:47)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3770)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:477)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:447)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3806)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3774)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3766)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:333)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:163)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:299)
        at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:206)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:114)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:394)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:341)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:199)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:186)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:111)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:100)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:214)
        at io.ballerina.projects.Package.getCompilation(Package.java:140)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:75)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:176)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:53)
```

**Affected Versions:**
slbeta3-SNAPSHOT

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #32271

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by gimantha — COMMENTED** [review]
Shall we enable the tests commented in this PR? https://github.com/ballerina-platform/ballerina-lang/pull/35248 @lasinicl @pcnfernando ?

File: tests/jballerina-unit-test/src/test/resources/test-src/narrowing/assignment_to_narrowed_vars_in_loops_negative.bal [hunk 1]
```diff
@@ -679,3 +679,42 @@ function f23() {
...
+
+function testInLoopAssignmentsWithQueryAction() returns error? {
+    (int|string)[] x = [1, 2];
+    int|string? p = 10;
+
+    if p is int|string {
+        check from var a in x
+            where p is int
+            do {
+                p = 12;
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/narrowing/assignment_to_narrowed_vars_in_loops_negative.bal:691]: error msg here "invalid attempt to assign a value to a variable narrowed outside the query action" is not correct; isn't it?

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/narrowing/assignment_to_narrowed_vars_in_loops_negative.bal:691]: Check other places as well

Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/resources/test-src/narrowing/assignment_to_narrowed_vars_in_loops_negative.bal:691]: `p` is defined outside the query and it's narrowed outside the query. So it should right?

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/narrowing/assignment_to_narrowed_vars_in_loops_negative.bal:691]: my bad. this is correct

**Review by pcnfernando — APPROVED** [review]

**Review by lochana-chathura — APPROVED** [review]

**Pull request merged**